mesonbuild/build.py

```diff
@@ -216,14 +216,14 @@ class InstallDir(HoldableObject):
 
 @dataclass(eq=False)
 class DepManifest:
-    version: str
+    version: T.Optional[str]
     license: T.List[str]
     license_files: T.List[T.Tuple[str, File]]
     subproject: str
 
     def to_json(self) -> T.Dict[str, T.Union[str, T.List[str]]]:
         return {
-            'version': self.version,
+            'version': self.version or "null",
```

Comment:
I don't think this is correct. We already default it to "undefined" and it shouldn't be possible to be T.Optional.

Reply:
Are you saying it's not possible for `self.version` to be `None`? This is `self.project_version` which is `T.Optional[str]`.

Reply:
Also, AIUI `"null"` is the right JSON representation of `None`

Reply:
`null` not `"null"`. And python will automatically convert it's `None` to `null`. It might be more appropriate to just change the return value annotation to `dict[str, str | list[str] | None]`

Reply:
Yes -- interpreter's `project_version` is set to `'undefined'` if you don't pass one in meson.build

intr.build.project_version is not used to create a DepManifest object -- only intr.project_version.

And intr.build.project_version can be None when build.Build() is constructed, but as soon as intr.func_project() is evaluated, it is updated to a string. So the sentinel "None" value is technically correct but not very representative of runtime.

Reply:
Hence, at a typing level, DepManifest cannot receive version=None and shouldn't bother handling it.

Reply:
In that case, I support just adding `assert self.version is not None, 'for mpy'`, which should be sufficient to make mypy happy, and documents the expectation.

Reply:
Why doesn't mypy like it to begin with?

It is this PR that changes the annotation of:
```diff
-self.version: str
+self.version: T.Optional[str]
```

Reply:
I suspect it's `validator=_project_version_validator()` that causes mypy to conclude that `self.project_version` is `str | None`. However, we also set `default='undefined'` in the same place, so I think the type checker is just wrong. I have fixed things up a bit.